classes/core/PKPEventServiceProvider.inc.php

```diff
@@ -78,6 +84,82 @@ public function boot()
             Event::subscribe($subscriber);
         }
     }
+
+    /**
+     * Determine if events and listeners should be automatically discovered.
+     *
+     * @return bool
+     */
+    public function shouldDiscoverEvents()
+    {
+        return true;
+    }
+
+    /**
+     * Get the discovered events for the application.
+     *
+     * @return array
+     */
+    protected function discoveredEvents()
+    {
+        return $this->shouldDiscoverEvents()
+            ? $this->discoverEvents()
+            : [];
+    }
+
+    /**
+     * Discover the events and listeners for the application.
+     *
+     * @return array
+     */
+    public function discoverEvents()
+    {
+
+        // Adapt classes naming convention
+        $discoverEvents = new class extends DiscoverEvents
+        {
+            /**
+             * @param SplFileInfo $file
+             * @param string $basePath base path of the application
+             * @return string listener's full class name
+             */
+            protected static function classFromFile(SplFileInfo $file, $basePath): string
+            {
+                $pathFromBase = trim(Str::replaceFirst($basePath, '', $file->getRealPath()), DIRECTORY_SEPARATOR);
+                $libPath = 'lib' . DIRECTORY_SEPARATOR . 'pkp' . DIRECTORY_SEPARATOR;
+                $namespace = Str::startsWith($pathFromBase, $libPath) ? 'PKP\\' : 'APP\\';
+
+                $path = $pathFromBase;
+                if ($namespace === 'PKP\\') $path = Str::replaceFirst($libPath, '', $path);
+                if (Str::startsWith($path, 'classes'))  $path = Str::replaceFirst('classes' . DIRECTORY_SEPARATOR, '', $path);
+                return $namespace . str_replace('/', '\\', Str::replaceLast('.inc.php', '', $path));
+            }
+        };
+
+        return collect($this->discoverEventsWithin())
+            ->reject(function ($directory) {
+                return ! is_dir($directory);
+            })
+            ->reduce(function ($discovered, $directory) use ($discoverEvents) {
+                return array_merge_recursive(
+                    $discovered,
+                    $discoverEvents::within($directory, base_path())
+                );
+            }, []);
+    }
+
+    /**
+     * Get the listener directories that should be used to discover events.
+     *
+     * @return array
+     */
+    protected function discoverEventsWithin()
```

Comment:
Does this scan happen on a per-request basis or does this use the Laravel event cache? (https://laravel.com/docs/8.x/events#event-discovery-in-production)

Reply:
No, it requires registration of at least Laravel's `FileServiceProvider`. Haven't dive deep into the code, but maybe `CacheServiceProvider` also. I'm not sure how slow the iteration [through listener's methods](https://github.com/laravel/framework/blob/cada3d21c7f127ba01bb949b2fa045b7faaedbca/src/Illuminate/Foundation/Events/DiscoverEvents.php#L55) with `ReflectionClass` is. Do you think event caching is something that should be implemented right away?

Reply:
On the second look, I can override Laravel's methods to use our file cache. Not necessary to integrate additional Laravel services. @asmecher, do you think it's worth an effort to do it now, before merging this PR?

Reply:
I'm content to start as it is, but we'll have to watch as we adopt the event toolset to see whether performance becomes an issue. Thanks!